posts/2018-11-02-semirings.md

```diff
@@ -0,0 +1,435 @@
+```
+
+## Higher kinded algebraic structures
+
+Some of you might have heard of the `Semigroupal` type class. 
+But why is it called that, and what is its relation to a `Semigroup`?
+Let's find out!
+
+First, let's have a look at `Semigroupal`:
+
+```scala
+@typeclass trait Semigroupal[F[_]] {
+  def product[A, B](fa: F[A], fb: F[B]): F[(A, B)]
+}
+```
+
+It seems to bear some similarity to `Semigroup`, we have two values which we somehow combine, and it also shares `Semigroup`s associativity requirement.
+
+So far so good, but the name `product` seems a bit weird.
+It makes sense given we combine the `A` and the `B` in a tuple, which is a product type, but if we're using products, maybe this isn't a generic `Semigroupal` but actually a multiplicative one? 
+Let's fix this and rename it!
+
+```tut:silent
+@typeclass trait MultiplicativeSemigroupal[F[_]] {
+  def product[A, B](fa: F[A], fb: F[B]): F[(A, B)]
+}
+```
+
+Next, let us have a look at what an additive `Semigroupal` might look like.
+Surely, the only thing we'd have to change is going from a product type to a sum type:
+
+```tut:silent
+@typeclass trait AdditiveSemigroupal[F[_]] {
+  def sum[A, B](fa: F[A], fb: F[B]): F[Either[A, B]]
+}
+```
+
+Pretty interesting so far, can we top this and add identities to make `Monoidal`s?
+Surely we can! For addition this should again be `Nothing` and `Unit` for multiplication:
+
+```tut:silent
+@typeclass trait AdditiveMonoidal[F[_]] extends AdditiveSemigroupal[F] {
+  def nothing: F[Nothing]
+}
+
+@typeclass trait MultiplicativeMonoidal[F[_]] extends MultiplicativeSemigroupal[F] {
+  def unit: F[Unit]
+}
+```
+
+So now we have these fancy type classes, but how are they actually useful?
+Well, I'm going to make the claim that these type classes already exist in cats today, just under different names.
+
+Let's first look at the `AdditiveMonoidal`.
+It is defined by two methods, `nothing` which returns an `F[Nothing]` and `sum` which takes an `F[A]` and an `F[B]` to create an `F[Either[A, B]]`.
+
+What type class in Cats could be similar?
+First, we'll look at the `sum` function and try to find a counterpart for `AdditiveSemigroupal`.
+Since we gave the lower kinded versions of these type classes symbolic operators, why don't we do the same thing for `AdditiveSemigroupal`?
+
+Since it is additive it should probably contain a `+` somewhere and it should also show that it's inside some context.
+
+Optimally it'd be something like `[+]`, but that's not a valid identifier so let's try `<+>` instead!
+
+```scala
+def <+>[A, B](fa: F[A], fb: F[B]): F[Either[A, B]]
+```
+
+Oh! The `<+>` function already exists in cats as an alias for `combineK` which can be found on `SemigroupK`, but it's sort of different, it takes two `F[A]`s and returns an `F[A]`, not quite what we have here.
+
+Or is it?
+These two functions are actually the same, and we can define them in terms of one another as long as we have an invariant functor:
```

Comment:
Did you intend to show the other direction too?

Reply:
No, I think just showing one should be fine, blog post is already plenty long as is :)